### PR TITLE
Add `strip_unneeded_headers_before_save`

### PR DESCRIPTION
**Pull request by shalvah-gs**
Following on from the proposal [on Slack](https://getsafe.slack.com/archives/C03DT3HD953/p1707487965863219), we want to get rid of unneeded headers in our VCR cassettes to keep the size down, and make them more maintainable.

This PR adds this functionality to the nxt_vcr_harness gem, so all that's needed is to call `NxtVcrHarness.strip_unneeded_headers_before_save`. The gem comes with a blacklist of common unneeded headers which will be removed.

Since this is a public gem, I've kept the list relatively light. Custom headers for services we use can be added in our gems, for instance:

```rb

NxtVcrHarness.strip_unneeded_headers_before_save do |headers_to_strip|
  headers_to_strip[:responses].concat [
    'X-Dixa-Rid', # Dixa
    'X-Mp-Request-Id', # Mixpanel
    'X-Sanity-Shard', # Sanity
    # Twilio
    'X-Shenanigans',
    'Twilio-Request-Duration',
  ]
end
```

For existing cassettes, I'm running a simple Rake task across our services (example in the README).

[BE-85](https://hellogetsafe.atlassian.net/browse/BE-85)

[BE-85]: https://hellogetsafe.atlassian.net/browse/BE-85?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/nxt_vcr_harness.rb [hunk 1]
```diff
@@ -66,5 +67,19 @@ def track_cassettes
     end
   end
 
-  module_function :enable_vcr_tag, :track_cassettes, :enable_vcr_cassette_helper, :track_cassettes_if
+  def strip_unneeded_headers_before_save
+    headers_to_strip = UnneededHeaders.default_headers_to_strip
+
+    yield headers_to_strip if block_given?
```
Comment (Thrashattack) [inline lib/nxt_vcr_harness.rb:73]: Can we have a method documentation listing all the headers being stripped by default, so when you hover the method you can have the documentation on the IDE and know wich one to exclude from the list if you wanna keep.

Comment (shalvah-gs) [reply lib/nxt_vcr_harness.rb:73]: For something to be used often, sure! But this is something you only add to your repo once, so autocomplete isn't much use. Additionally, adding a manual list would double the size of the file, and get out of date easily.

As mentioned in the README, if you want to see the removed headers, you can just call `UnneededHeaders.default_headers_to_strip`

**Review by TylerWitt — COMMENTED** [review]
Maybe it's more clear to choose a list of headers to keep instead? I doubt many of them are critical for VCR checking.

**Comment by shalvah-gs** [comment]
> Maybe it's more clear to choose a list of headers to keep instead?

We went with this, because it's the approach of least surprise to the end user. It's  safer to remove known redundant headers (this already gets us huge savings) than only keep known headers. Imagine adding a library (eg Calendly gem), and your tests keep failing because it depends on a header we don't keep. That would be fun to debug!

**Comment by TylerWitt** [comment]
> > Maybe it's more clear to choose a list of headers to keep instead?
> 
> We went with this, because it's the approach of least surprise to the end user.

My suggestion missed that this was a public gem--so I agree (though I'd argue this should be a private one)

>  Imagine adding a library (eg Calendly gem), and your tests keep failing because it depends on a header we don't keep. That would be fun to debug!

These headers are only used for matching the cassette right? It shouldn't have any functional implications as long as the headers are matching.

**Comment by shalvah-gs** [comment]
> These headers are only used for matching the cassette right? It shouldn't have any functional implications as long as the headers are matching.

It can. When you make an API call, it gets to the low-level client (eg Typhoeus), and just before it fires the request, VCR takes over and returns the cassette as if that's the response. If you're trying to paginate over a collection, if the pagination headers are missing, you'll have wrong values.

(Btw, in most cases, headers aren't used for matching. Only URL and path.)

**Review by TylerWitt — APPROVED** [review]

**Pull request merged**